Secure Issue 40593: User not prompted for registration survey after password setting

labkey-bpatel opened the pull request:
#### Rationale
NeedSurvey property was determined using lastLogin being null, but it is no longer the case. Also, cds.appAction is not a mutating action so is unable to set the NeedSurvey value.

#### Changes
- Create a new MutatingApiAction that sets NeedSurvey value to true after newly registered user sets the password.

labkey-bpatel committed:
Set NeedSurvey param when creating an account.

labkey-bpatel committed:
Move setting of NeedSurvey prop to a new Mutating api action.
Include this new action in the success call for setPasswordAPI.

XingY commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -226,16 +226,33 @@ public void setUserProperties(JSONObject userProperties)
 
     private static final String ANALYTICS_USER_GROUP = "Active CAVD Member";
 
+    @RequiresNoPermission
```
should require read permission for updates

labkey-bpatel replied:
Done!

XingY commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -226,16 +226,33 @@ public void setUserProperties(JSONObject userProperties)
             Container c = getContainer();
             StudyService sss = StudyService.get();
             boolean hasStudy = null != sss && null != sss.getStudy(c);
+            boolean needSurvey = Boolean.parseBoolean(getViewContext().getActionURL().getParameter("NeedSurvey"));
```
I don't think this is needed with the ajax, right?

labkey-bpatel replied:
I had to add this specifically because  line 266, CDSManager.get().isNeedSurvey(getUser()) was returning false. Didn't think it was needed either, esp. since this code is executed as part of success callback after setting NeedSurvey to true.

labkey-bpatel replied:
This is now fixed by OntologyManager to get NeedSurvey value as we discussed.

XingY commented on webapp/frontPage/js/modal.js:
```diff
@@ -395,14 +402,25 @@ define(['jquery', 'magnific', 'util'], function($, magnific, util) {
+                'X-LABKEY-CSRF': LABKEY.CSRF
+              }
+            }).success(function() {
+              window.location = LABKEY.ActionURL.buildURL("cds", "app.view", null, {NeedSurvey:true}); // set password should log user in automatically
```
instead of setting NeedSurvey to true, which would trigger another redirect in controller, could instead use param "survey=true", which is actually what prompts survey page.

labkey-bpatel replied:
done!

labkey-bpatel committed:
Code Review changes:
- Remove 'NeedSurvey' from param, use 'survey' instead.
- Get NeedSurvey value using OntologyManager.
- Combine code for guest user and registered user requiring survey.

labkey-bpatel committed:
Code Review changes:
- Require ReadPermission.

XingY reviewed (approved):
looks great

XingY commented:
After buddy testing, I wonder if it's better to still let user proceed to app with "Unable to set NeedSurvey property to true" error. We probably shouldn't stop them from accessing site if they can't fill out survey due to tech issue. Admin can tell user wasn't able to complete survey from site logs. Also if they refresh the page, they will be redirected to the app anyways, since they are logged in. But if they are presented with this error, they might try to click "Create Account" button again, which is gonna say "unable to... account already created.." error, which is confusing.

But otherwise, buddy testing looks good.

labkey-bpatel committed:
Bug fix : Navigate users to app view even if there's an error setting needSurvey property.

labkey-bpatel commented:
@XingY - User can now access the site if they can't fill out survey due to tech issue.

XingY commented on webapp/frontPage/js/modal.js:
```diff
@@ -395,14 +402,26 @@ define(['jquery', 'magnific', 'util'], function($, magnific, util) {
+            }).error(function(e) {
+              var errorMsg = 'Unable to set NeedSurvey property to true';
+              createAccountError(e, errorMsg);
+              window.location = LABKEY.ActionURL.buildURL("cds", "app.view");
```
maybe put redirect in a setTimeout call so user can briefly see the error (or maybe skip error at all since user won't make any sense of the msg)

labkey-bpatel replied:
Good point! i've added error to console.log instead for dev troubleshooting.

XingY replied:
Good call. Another (one last) minor change: use console.error instead of console.log.

labkey-bpatel replied:
Done!

labkey-bpatel committed:
Remove error, not useful to the user; instead add to console.log() for dev troubleshooting.

labkey-bpatel committed:
Change to console.error from console.log.
